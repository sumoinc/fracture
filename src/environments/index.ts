export * from "./environment";
export * from "./region";
