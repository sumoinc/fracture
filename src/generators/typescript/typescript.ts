import { SourceCode } from "projen";
import { FractureComponent } from "../../core/component";
import { Fracture } from "../../core/fracture";

export class TypeScript extends FractureComponent {
  constructor(fracture: Fracture) {
    super(fracture);
  }

  public preSynthesize() {
    /**
     * Build a TypeScript file and make sure it's marked as being managed by
     * projen.
     *
     * @param name
     * @returns {SourceCode}
     */
    const typeScriptFile = (name: string): SourceCode => {
      const f = new SourceCode(
        this.project,
        `${this.fracture.gendir}/types/${name}`
      );
      f.line("// " + f.marker);
      f.line("\n");
      return f;
    };

    // create index file
    const indexFile = typeScriptFile("index.ts");

    /**
     * Generate types for each entity.
     */
    this.fracture.entities.forEach((e) => {
      // add to index for export
      indexFile.line(`export * from "./${e.name}";`);

      // generate types for this entity
      const tsFile = typeScriptFile(`${e.name}.ts`);

      // generate helpful comments
      tsFile.line(`/**`);
      e.comment.forEach((c) => tsFile.line(` * ${c}`));
      tsFile.line(` */`);

      // build interface
      tsFile.open(`export interface ${e.name} {`);
      e.attributes.forEach((a) => {
        // generate helpful comments
        tsFile.line(`/**`);
        a.comment.forEach((c) => tsFile.line(` * ${c}`));
        tsFile.line(` */`);
        // console.log(a);

        // write type definition
        const q = a.isRequired ? "" : "?";
        tsFile.line(`${a.name}${q}: ${a.typeScriptType};`);
      });
      tsFile.close(`}`);
      tsFile.line("\n");
    });
    // close up index file
    indexFile.line("\n");
  }
}
