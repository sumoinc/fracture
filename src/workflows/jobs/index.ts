export * from "./build-job";
export * from "./deploy-job";
