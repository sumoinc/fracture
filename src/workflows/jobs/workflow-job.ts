import { Component } from "projen";
import { Job } from "projen/lib/github/workflows-model";
import { NodeProject } from "projen/lib/javascript";

export interface WorkflowJobOptions {
  /**
   * Job's Id
   */
  readonly jobId: string;
  /**
   * Job's name
   */
  readonly name: string;
}

export class WorkflowJob extends Component {
  /**
   * Build Job's Id for this workflow.
   */
  public readonly jobId: string;
  /**
   * Job's name
   */
  public readonly name: string;

  constructor(
    public readonly project: NodeProject,
    options: WorkflowJobOptions
  ) {
    super(project);

    //defaults
    this.jobId = options.jobId;
    this.name = options.name;
  }

  render = (): Job => {
    throw new Error("Not Implemented");
  };
}
