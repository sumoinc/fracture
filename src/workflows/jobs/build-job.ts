import { Job, JobPermission } from "projen/lib/github/workflows-model";
import { SetOptional } from "type-fest";
import { WorkflowJob, WorkflowJobOptions } from "./workflow-job";
import { renderBuildSteps } from "../steps/build-steps";
import { renderUploadArtifactSteps } from "../steps/upload-artifact-steps";
import { Workflow } from "../workflow";

export interface BuildJobOptions
  extends SetOptional<WorkflowJobOptions, "jobId" | "name"> {}

export class BuildJob extends WorkflowJob {
  constructor(
    public readonly workflow: Workflow,
    options: BuildJobOptions = {}
  ) {
    super(workflow, {
      jobId: "build",
      name: "Build",
      ...options,
    });
  }

  render = (): Job => {
    return {
      name: "Build",
      runsOn: this.workflow.defaultRunners,
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.WRITE,
      },
      steps: [
        ...renderBuildSteps(this.workflow),
        ...renderUploadArtifactSteps(this),
      ],
    };
  };
}
