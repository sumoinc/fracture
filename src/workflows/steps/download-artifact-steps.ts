import crypto from "crypto";
import { JobStep } from "projen/lib/github/workflows-model";
import { WorkflowJob } from "../jobs/workflow-job";
import { PERMISSION_BACKUP_FILE } from "../workflow";

export const renderDownloadArtifactSteps = (
  job: WorkflowJob
): Array<JobStep> => {
  const steps: Array<JobStep> = [];

  // dedupe and loop
  [...new Set(job.artifactDirectories)].forEach((artifactDirectory) => {
    // hash the name of the artifact directory to avoid collisions
    let hash = crypto
      .createHash("md5")
      .update(artifactDirectory)
      .digest("hex")
      .substring(0, 8);

    // add artifact upload steps
    steps.push(
      {
        name: `Download build artifacts for "${artifactDirectory}"`,
        uses: "actions/download-artifact@v3",
        with: {
          name: `build-artifact-${hash}`,
          path: artifactDirectory,
        },
      },
      {
        name: `Restore build artifact permissions for "${artifactDirectory}"`,
        continueOnError: true,
        run: `setfacl --restore=${PERMISSION_BACKUP_FILE}`,
        workingDirectory: artifactDirectory,
      }
    );
  });

  return steps;
};
