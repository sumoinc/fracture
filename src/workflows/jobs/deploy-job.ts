import { paramCase } from "change-case";
import { Job, JobPermission, JobStep } from "projen/lib/github/workflows-model";
import { TypeScriptProject } from "projen/lib/typescript";
import { SetOptional } from "type-fest";
import { WorkflowJob, WorkflowJobOptions } from "./workflow-job";
import { Environment } from "../../environments";
import { renderDeploySteps } from "../steps/deploy-steps";
import { renderDownloadArtifactSteps } from "../steps/download-artifact-steps";
import { renderSetupNodeSteps } from "../steps/setup-node-steps";
import { Workflow } from "../workflow";

export interface DeployJobOptions
  extends SetOptional<WorkflowJobOptions, "jobId" | "name"> {
  /**
   * How to identify the app being deployed
   *
   */
  readonly appName: string;

  /**
   * The branch pattern this deployment is targeting.
   *
   * @default main
   */
  readonly branchPrefix?: string;

  /**
   * Jobs that must complete before this one may run
   *
   * @default - the build job
   */
  readonly needs?: Array<string>;

  /**
   * Step definition(s) that conduct the deployment
   */
  readonly deploySteps: Array<JobStep>;

  /**
   * Environment deploying to.
   */
  readonly environment: Environment;

  /**
   * Directory that was preserved for this project in order to deploy
   * properly in deploy step.
   */
  readonly artifactsDirectory: string;
}

export class DeployJob extends WorkflowJob {
  /**
   * The branch pattern this deployment is targeting.
   *
   * @default main
   */
  public readonly branchPrefix: string;

  /**
   * Jobs that must complete before this one may run
   *
   * @default - the build job
   */
  public readonly needs: Array<string>;

  /**
   * Step definition(s) that conduct the deployment
   */
  public readonly deploySteps: Array<JobStep>;

  /**
   * Environment deploying to.
   */
  public readonly environment: Environment;

  /**
   * Directory that was preserved for this project in order to deploy
   * properly in deploy step.
   */
  public readonly artifactsDirectory: string;

  /**
   * Workflow for deployment
   */
  readonly workflow: Workflow;

  constructor(
    public readonly project: TypeScriptProject,
    options: DeployJobOptions
  ) {
    const branchPrefix = options.branchPrefix ?? "main";
    const name = `Deploy ${options.appName} to ${options.environment.name}`;
    const jobId = options.jobId ?? paramCase(name);

    super(project, {
      jobId,
      name,
      ...options,
    });

    // defaults
    this.workflow = Workflow.deploy(project);
    this.branchPrefix = branchPrefix;
    this.needs = [this.workflow.buildJob.jobId, ...(options.needs ?? [])];
    this.deploySteps = options.deploySteps;
    this.environment = options.environment;
    this.artifactsDirectory = options.artifactsDirectory;

    // add this branch pattern as a trigger for the workflow
    this.workflow.addPushTrigger(this.branchPrefix);

    // make sure build step will store the artifacts we need
    this.workflow.buildJob.artifactDirectories.push(this.artifactsDirectory);
  }

  dependsOn = (job: WorkflowJob): void => {
    this.needs.push(job.jobId);
  };

  render = (): Job => {
    return {
      name: this.name,
      runsOn: this.workflow.defaultRunners,
      needs: this.needs,
      concurrency: this.jobId,
      if: `startsWith( github.ref, 'refs/heads/${this.branchPrefix}' )`,
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.READ,
        ...(this.environment.authProviderType === "AWS_GITHUB_OIDC"
          ? {
              idToken: JobPermission.WRITE,
            }
          : {}),
      },
      steps: [
        ...renderDownloadArtifactSteps(this),
        ...renderSetupNodeSteps(this.workflow, {
          packageCache: false,
          setupPnpm: false,
        }),
        ...renderDeploySteps(this),
      ],
    };
  };
}
