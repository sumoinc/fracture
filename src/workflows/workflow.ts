import { Component } from "projen";
import { GitHub, GithubWorkflow } from "projen/lib/github";
import { TypeScriptProject } from "projen/lib/typescript";
import { BuildJob } from "./jobs/build-job";
import { DeployJob, DeployJobOptions } from "./jobs/deploy-job";
import { WorkflowJob } from "./jobs/workflow-job";

/**
 * Name of the permission back up file to include in the build artifact
 * to work around a GitHub Action bug that does not preserve file mode
 * permissions across upload and download actions.
 *
 * See {@link https://github.com/actions/upload-artifact/issues/38}
 */
export const PERMISSION_BACKUP_FILE = "permissions-backup.acl";

export interface WorkflowOptions {
  /**
   * Name of the file (e.g. "deploy" becomes "deploy.yml").
   *
   * @default workflow
   */
  readonly name?: string;

  /**
   * The node version to use in workflows.
   *
   * @default latest
   */
  readonly nodeVersion?: string;

  /**
   * Enable Node.js package cache in GitHub workflows.
   *
   * @default true
   */
  readonly packageCache?: boolean;

  /**
   * The version of PNPM to use if using PNPM as a package manager.
   *
   * @default "8"
   */
  readonly pnpmVersion?: string;

  /**
   * Automatically update files modified during builds to pull-request branches.
   * This means that any files synthesized by projen or e.g. test snapshots will
   * always be up-to-date before a PR is merged.
   *
   * Implies that PR builds do not have anti-tamper checks.
   *
   * This is enabled by default only if `githubTokenSecret` is set. Otherwise it
   * is disabled, which implies that file changes that happen during build will
   * not be pushed back to the branch.
   *
   * @default false
   */
  readonly mutableBuild?: boolean;

  /**
   * What runners should this workflow use by default?
   *
   * @default ["ubuntu-latest"]
   */
  readonly defaultRunners?: Array<string>;
}

export class Workflow extends Component {
  /**
   * Name of the file (e.g. "deploy" becomes "deploy.yml").
   */
  public readonly name: string;

  /**
   * The node version to use in workflows.
   *
   * @default latest
   */
  public readonly nodeVersion: string;

  /**
   * Enable Node.js package cache in GitHub workflows.
   *
   * @default true
   */
  public readonly packageCache: boolean;

  /**
   * Branch or tag name.
   * @default - the default branch is implicitly used
   */
  public readonly ref?: string;

  /**
   * The repository (owner/repo) to use.
   * @default - the default repository is implicitly used
   */
  public readonly repository?: string;

  /**
   * Whether to download files from Git LFS for this workflow
   *
   * @default - Use the setting on the corresponding GitHub project
   */
  public readonly downloadLfs?: boolean;

  /**
   * The version of PNPM to use if using PNPM as a package manager.
   *
   * @default "8"
   */
  public readonly pnpmVersion: string;

  /**
   * Automatically update files modified during builds to pull-request branches.
   * This means that any files synthesized by projen or e.g. test snapshots will
   * always be up-to-date before a PR is merged.
   *
   * Implies that PR builds do not have anti-tamper checks.
   *
   * This is enabled by default only if `githubTokenSecret` is set. Otherwise it
   * is disabled, which implies that file changes that happen during build will
   * not be pushed back to the branch.
   *
   * @default false
   */
  readonly mutableBuild: boolean;

  /**
   * What runners should this workflow use by default?
   *
   * @default ["ubuntu-latest"]
   */
  readonly defaultRunners: Array<string>;

  /**
   * Instance of github configuration for the project
   */
  public readonly github: GitHub;

  /**
   * Github workflow Workflow
   */
  public readonly workflow: GithubWorkflow;

  /**
   * Build Job for the workflow.
   */
  public readonly buildJob: BuildJob;

  /**
   * Other Jobs in this workflow
   */
  public readonly otherJobs: Array<WorkflowJob> = [];

  constructor(
    public readonly project: TypeScriptProject,
    options: WorkflowOptions
  ) {
    super(project);

    const github = GitHub.of(project);

    if (!github) {
      throw new Error(
        "Workflows are currently only supported for GitHub projects"
      );
    }

    // defaults
    this.name = options.name ?? "workflow";
    this.nodeVersion = options.nodeVersion ?? "latest";
    this.packageCache = options.packageCache ?? true;
    this.pnpmVersion = options.pnpmVersion ?? "8";
    this.mutableBuild = options.mutableBuild ?? false;
    this.defaultRunners = options.defaultRunners ?? ["ubuntu-latest"];

    // local
    this.github = github;
    this.workflow = new GithubWorkflow(github, this.name);

    // render the build steps
    this.buildJob = new BuildJob(this);
  }

  addDeployJob(options: DeployJobOptions): DeployJob {
    const job = new DeployJob(this, { ...options });
    this.otherJobs.push(job);
    return job;
  }

  preSynthesize(): void {
    // add the build job to the workflow
    this.workflow.addJob(this.buildJob.jobId, this.buildJob.render());
    // add other jobs
    this.otherJobs.forEach((job) => {
      this.workflow.addJob(job.jobId, job.render());
    });
  }
}
