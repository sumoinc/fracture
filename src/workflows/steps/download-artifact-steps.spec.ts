import { JobStep } from "projen/lib/github/workflows-model";
import { TypeScriptProject } from "projen/lib/typescript";
import { renderDownloadArtifactSteps } from "./download-artifact-steps";
import { Environment } from "../../environments";
import { DeployJob } from "../jobs/deploy-job";
import { Workflow } from "../workflow";

const deploySteps: Array<JobStep> = [
  {
    name: "Say foo",
    run: "echo 'foo'",
  },
];

describe("success conditions", () => {
  test("Smoke test", () => {
    const project = new TypeScriptProject({
      name: "my-project",
      defaultReleaseBranch: "main",
    });
    const workflow = new Workflow(project, {
      name: "my-workflow",
    });
    const environment = new Environment(project, {
      name: "my-environment",
    });
    const deployJob = new DeployJob(workflow, { deploySteps, environment });
    const artifactSteps = renderDownloadArtifactSteps(deployJob);
    expect(artifactSteps).toBeTruthy();
    expect(artifactSteps).toMatchSnapshot();
    // console.log(artifactSteps);
  });

  test("With Deplopy Job", () => {
    const project = new TypeScriptProject({
      name: "my-project",
      defaultReleaseBranch: "main",
    });
    const workflow = new Workflow(project, {
      name: "my-workflow",
    });
    const environment = new Environment(project, {
      name: "my-environment",
    });
    const deployJob = new DeployJob(workflow, {
      artifactDirectories: ["some/other/folder/cdk.out"],
      deploySteps,
      environment,
    });
    const artifactSteps = renderDownloadArtifactSteps(deployJob);
    expect(artifactSteps).toBeTruthy();
    expect(artifactSteps).toMatchSnapshot();
    // console.log(JSON.stringify(artifactSteps, null, 2));
  });
});
