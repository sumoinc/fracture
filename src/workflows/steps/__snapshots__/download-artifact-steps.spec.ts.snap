// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`success conditions Smoke test 1`] = `
Array [
  Object {
    "name": "Download build artifacts for \\"foo\\"",
    "uses": "actions/download-artifact@v3",
    "with": Object {
      "name": "build-artifact-acbd18db",
      "path": "foo",
    },
  },
  Object {
    "continueOnError": true,
    "name": "Restore build artifact permissions for \\"foo\\"",
    "run": "setfacl --restore=permissions-backup.acl",
    "workingDirectory": "foo",
  },
]
`;

exports[`success conditions With Deplopy Job 1`] = `
Array [
  Object {
    "name": "Download build artifacts for \\"foo\\"",
    "uses": "actions/download-artifact@v3",
    "with": Object {
      "name": "build-artifact-acbd18db",
      "path": "foo",
    },
  },
  Object {
    "continueOnError": true,
    "name": "Restore build artifact permissions for \\"foo\\"",
    "run": "setfacl --restore=permissions-backup.acl",
    "workingDirectory": "foo",
  },
]
`;
